src/execution_plan/execution_plan_modify.c

```diff
@@ -68,63 +52,33 @@ void _OpBase_RemoveNode(OpBase *parent, OpBase *child) {
-	 * at this point of execution, should include all
-	 * previously modified entities (up the execution plan). */
-	const char **seen = array_new(const char *, 0);
-
-	uint modifies_count = array_len(root->modifies);
-	/* Append current op modified entities. */
-	for(uint i = 0; i < modifies_count; i++) {
-		seen = array_append(seen, root->modifies[i]);
-		// printf("%u\n", root->modifies[i]);
-	}
+bool _ExecutionPlan_LocateReferences(OpBase *root, OpBase **op, const OpBase *recurse_limit,
+									 rax *references_to_resolve, bool subtree_is_dependent) {
+	if(root == recurse_limit) return false; // Don't traverse into earlier ExecutionPlan scopes.
 
-	/* Traverse execution plan, upwards. */
+	bool refs_resolved = false;
 	for(int i = 0; i < root->childCount; i++) {
-		const char **saw = _ExecutionPlan_LocateReferences(root->children[i], op, references);
-
-		/* Quick return if op was located. */
-		if(*op) {
-			array_free(saw);
-			return seen;
-		}
-
-		uint saw_count = array_len(saw);
-		/* Append current op modified entities. */
-		for(uint i = 0; i < saw_count; i++) {
-			seen = array_append(seen, saw[i]);
-		}
-		array_free(saw);
+		/* Recurse to each child to resolve references. The final argument indicates whether references
+		 * have already been resolved - if so, the child cannot be the fully-resolving op. */
+		refs_resolved |= _ExecutionPlan_LocateReferences(root->children[i], op, recurse_limit,
+														 references_to_resolve, refs_resolved | subtree_is_dependent);
+		/* Return early if the resolving op has been set. */
+		if(*op) return true;
 	}
 
-	// Sort the 'seen' array and remove duplicate entries.
-	// This is necessary to properly intersect 'seen' with the 'references' rax.
-	_uniqueArray(seen);
-
-	uint seen_count = array_len(seen);
-
-	/* See if all references have been resolved. */
-	uint match = raxSize(references);
-
-	for(uint i = 0; i < seen_count; i++) {
-		// Too many unmatched references.
-		if(match > (seen_count - i)) break;
-		const char *seen_id = seen[i];
-
-		if(raxFind(references, (unsigned char *)seen_id, strlen(seen_id)) != raxNotFound) {
-			match--;
-			// All references have been resolved.
-			if(match == 0) {
-				*op = root;
-				break;
-			}
-		}
+	uint modifies_count = array_len(root->modifies);
+	for(uint i = 0; i < modifies_count; i++) {
+		const char *ref = root->modifies[i];
+		// Attempt to remove the current op's references, marking whether any removal was succesful.
+		refs_resolved |= raxRemove(references_to_resolve, (unsigned char *)ref, strlen(ref), NULL);
 	}
 
-	// if(!match) *op = root; // TODO might be necessary for post-WITH segments
-	return seen;
+	/* Set the op if all references have been resolved in this subtree.
+	 * If subtree_is_dependent is true, references have been resolved outside of this subtree, and the
+	 * filter op must be placed higher in the tree. */
+	if(!subtree_is_dependent && raxSize(references_to_resolve) == 0) *op = root;
```

Comment:
If `subtree_is_dependent` is ever set as a result of a sub match, when we will finally resolve all references, we'll recurse back to the point where `subtree_is_dependent` was first set (some sort of an intersection operation, e.g. cartesian product).
This is just a suggestion but we can replace `subtree_is_dependent` with a `OpBase *branch` an operation where the first diverge occurred, once all references been resolved (via multiple branches) simply set `op` to `branch` and return true.
No need to recurse backwards and perform logic outside of loop.

On a second thought we don't need an "extra" argument we can "ride" on `op` and return true only when both `op` is set and `raxSize(references_to_resolve) == 0`